### ISSUE TITLE
sql: overload resolution is inconsistent for functions that take numeric constants

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#14652]
While investigating #10407, I noticed some inconsistent behavior in our function overload resolver.

Currently, you can't take the `sqrt` of an integer literal without a type error:

```
root@:26257/test> select sqrt(3);
pq: unknown signature: sqrt(int)
```

However, by carefully constructing a query that uses tuple comparison (and possibly in other situations as well), you can:

```
root@:26257/test> SELECT (abs(10.0), 1) = (sqrt(100), 1);
+--------------------------------+
| (abs(10.0), 1) = (sqrt(100), 1) |
+--------------------------------+
| true                           |
+--------------------------------+
(1 row)
```

Note that flipping the operands here ends up failing:

```
root@:26257/test> SELECT (sqrt(100), 1) = (pow(1, 10.0), 1);
pq: unknown signature: sqrt(int)
```

I'm planning on fixing this tuple comparison inconsistency as part of #10407, but I wanted to document this for posterity. I'd also like to know whether it's intended that `sqrt(10)` doesn't resolve to either of the overloads for `sqrt` - at first glance, that seems wrong to me. Ideally, I think the function overload machinery should be happy to upcast the integer literal to either the decimal or the float - I think the trouble is that since there's two possibilities, it doesn't pick one early enough and ends up trying and failing to find an overload that accepts the "best" type for an integer constant: an integer.

@nvanbenschoten, what do you think?

**Comment by nvanbenschoten** [comment]
@jordanlewis thanks for resolving the tuple inconsistency!

The `sqrt(10)` is a little more subtle, and has to do with how overload resolution interacts with untyped constant literals. Your hunch is correct in that right now we fail to find an overload that fits the "best" type for the integer constant, so we report an ambiguity between the other two possibilities. We already treat a `Constant's` `AvailableTypes` as a list ordered by type preference, so ideally we'd  recognize that decimal is the best available type that's applicable for the constant and resolve it as that, avoiding the candidate ambiguity. I seem to recall complications with trying to add in logic like that to `typeCheckOverloadedExprs`, but I can't remember off the top of my head what those were. If you want to look into improving this I do think it would be worthwhile.

**Comment by dianasaur323** [comment]
I marked this as a 1.1 milestone, but please move forward if you think we should get to this earlier.

**Comment by knz** [comment]
We have addressed this in the 19.1 lifecycle - closing.
